File: iso27001-nestjs/src/domain/users/user.service.ts
import { Inject, Injectable, ConflictException } from '@nestjs/common';
import { EventEmitter2 } from '@nestjs/event-emitter';
import * as crypto from 'crypto';
import * as bcrypt from 'bcrypt';
import { User } from './user.entity';
import { IUserRepository, USER_REPOSITORY } from './user-repository.interface';
import { CreateUserDto } from './dto/create-user.dto';
import { UpdateUserDto } from './dto/update-user.dto';
import { UserCreatedEvent } from './events/user-created.event';

const BCRYPT_ROUNDS = 12; // A.10: bcrypt cost ≥ 12

@Injectable()
export class UserService {
  constructor(
    @Inject(USER_REPOSITORY)
    private readonly repo: IUserRepository,
    private readonly events: EventEmitter2,
  ) {}

  async create(dto: CreateUserDto): Promise<User> {
    if (await this.repo.existsByEmail(dto.email)) {
      throw new ConflictException('Email already exists');
    }

    // A.10: bcrypt password hashing (cost 12)
    const hashedPassword = await bcrypt.hash(dto.password, BCRYPT_ROUNDS);

    const user = new User();
    user.email = dto.email;
    user.hashedPassword = hashedPassword;
    user.fullName = dto.fullName ?? null;
    user.role = 'viewer';
    user.isActive = true;

    const saved = await this.repo.save(user);

    // A.12: Emit domain event for audit trail — never log raw email
    const emailHash = crypto.createHash('sha256').update(saved.email).digest('hex');
    this.events.emit('user.created', new UserCreatedEvent(saved.id, emailHash, saved.role));

    return saved;
  }

  async findAll(skip: number, limit: number): Promise<User[]> {
    return this.repo.findAll(skip, limit);
  }

  async findById(id: string): Promise<User | null> {
    return this.repo.findById(id);
  }

  async findByEmail(email: string): Promise<User | null> {
    return this.repo.findByEmail(email);
  }

  async update(user: User, dto: UpdateUserDto): Promise<User> {
    if (dto.email && dto.email !== user.email) {
      if (await this.repo.existsByEmail(dto.email)) {
        throw new ConflictException('Email already exists');
      }
      user.email = dto.email;
    }
    if (dto.fullName !== undefined) {
      user.fullName = dto.fullName;
    }
    return this.repo.save(user);
  }

  async delete(id: string): Promise<void> {
    return this.repo.softDelete(id);
  }

  /** A.9: Verify password with constant-time bcrypt comparison */
  async verifyPassword(plaintext: string, hash: string): Promise<boolean> {
    return bcrypt.compare(plaintext, hash);
  }
}

File: iso27001-springboot/src/main/java/com/iso27001/api/domain/users/UserService.java
package com.iso27001.api.domain.users;

import com.iso27001.api.domain.users.events.UserCreatedEvent;
import org.springframework.context.ApplicationEventPublisher;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.HexFormat;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

/**
 * A.9, A.10 — Core user domain service.
 * - Passwords hashed with bcrypt cost 12 (A.10)
 * - Domain events emitted with email hash, never raw PII (A.12)
 */
@Service
public class UserService {

    private final UserRepository repo;
    private final PasswordEncoder passwordEncoder;
    private final ApplicationEventPublisher events;

    public UserService(UserRepository repo, PasswordEncoder passwordEncoder, ApplicationEventPublisher events) {
        this.repo = repo;
        this.passwordEncoder = passwordEncoder;
        this.events = events;
    }

    @Transactional
    public User create(String email, String password, String fullName, UserRole role) {
        if (repo.existsByEmail(email)) {
            throw new IllegalArgumentException("Email already registered");
        }
        User user = new User();
        user.setEmail(email);
        user.setHashedPassword(passwordEncoder.encode(password));
        user.setFullName(fullName);
        user.setRole(role != null ? role : UserRole.VIEWER);
        user.setActive(true);

        User saved = repo.save(user);

        // A.12 — emit domain event with hashed email, never raw
        String emailHash = sha256(saved.getEmail());
        events.publishEvent(new UserCreatedEvent(saved.getId().toString(), emailHash, saved.getRole().name()));

        return saved;
    }

    public boolean verifyPassword(String plaintext, String hash) {
        // A.10 — constant-time comparison via bcrypt
        return passwordEncoder.matches(plaintext, hash);
    }

    @Transactional(readOnly = true)
    public Optional<User> findById(UUID id) {
        return repo.findById(id);
    }

    @Transactional(readOnly = true)
    public Optional<User> findByEmail(String email) {
        return repo.findByEmail(email);
    }

    @Transactional(readOnly = true)
    public List<User> findAll(int skip, int limit) {
        return repo.findAll(skip, limit);
    }

    @Transactional
    public User update(User user) {
        return repo.save(user);
    }

    @Transactional
    public void softDelete(UUID id) {
        repo.softDelete(id);
    }

    private static String sha256(String input) {
        try {
            byte[] digest = MessageDigest.getInstance("SHA-256")
                .digest(input.getBytes(StandardCharsets.UTF_8));
            return HexFormat.of().formatHex(digest);
        } catch (NoSuchAlgorithmException e) {
            throw new IllegalStateException("SHA-256 not available", e);
        }
    }
}
